Add JSO-style callback support, tests, and docs

arnavk23 opened the pull request:
- NLPModelsIpopt.jl — Try JSO-style callback (nlp, solver, stats) first, fallback to Ipopt-style callback.
- runtests.jl — Added a unit test verifying the JSO-style callback is accepted and can stop the solver.
- tutorial.md — Added documentation and a short example for the JSO-style callback.

Closes #104

arnavk23 committed:
Add JSO-style callback support, tests, and docs

arnavk23 committed:
Restructuring runtests

tmigot commented on test/runtests.jl:
```diff
@@ -65,6 +65,20 @@ end
+  stats = ipopt(nlp, tol = 1e-12, callback = jso_callback, print_level = 0)
+  @test stats.status == :user
+  @test stats.solver_specific[:internal_msg] == :User_Requested_Stop
+  @test stats.iter == 5
```
Remove comments and encapsulate this in a @testset  env. Remove println from tests.

arnavk23 committed:
Update docs/src/tutorial.md

Co-authored-by: Tangi Migot <tangi.migot@gmail.com>

arnavk23 committed:
Support JSO callback signature and Ipopt short/full signatures with explicit callback_style

- Introduce `callback_style` keyword with :auto (default), :jso, :ipopt_short, :ipopt_full
- In :auto, prefer JSO-style, then full Ipopt, then short Ipopt 3-arg
- Normalize callback return to Bool; `nothing` => true
- Update tests:
  * Add JSO callback tests (stop at 5; problem/nlp access; use solver/nlp)
  * Add short Ipopt 3-arg tests and force with callback_style=:ipopt_short
  * Fix obj_value type assertion
- All tests now pass locally

arnavk23 committed:
comment changes

arnavk23 commented:
@dpo Saw some of your commits recently on this pr. I think we can also go about this pr and merge after review.
